airbyte-integrations/connectors/source-freshdesk/source_freshdesk/api.py

```diff
@@ -0,0 +1,225 @@
+
+    @staticmethod
+    def _parse_and_handle_errors(req):
+        try:
+            j = req.json()
+        except ValueError:
+            j = {}
+
+        error_message = "Freshdesk Request Failed"
+        if "errors" in j:
+            error_message = "{}: {}".format(j.get("description"), j.get("errors"))
+        elif "message" in j:
+            error_message = j["message"]
+
+        if req.status_code == 400:
+            raise FreshdeskBadRequest(error_message)
+        elif req.status_code == 401:
+            raise FreshdeskUnauthorized(error_message)
+        elif req.status_code == 403:
+            raise FreshdeskAccessDenied(error_message)
+        elif req.status_code == 404:
+            raise FreshdeskNotFound(error_message)
+        elif req.status_code == 429:
+            raise FreshdeskRateLimited(
+                "429 Rate Limit Exceeded: API rate-limit has been reached until {} seconds. See "
+                "http://freshdesk.com/api#ratelimit".format(req.headers.get("Retry-After"))
+            )
+        elif 500 < req.status_code < 600:
+            raise FreshdeskServerError("{}: Server Error".format(req.status_code))
+
+        # Catch any other errors
+        try:
+            req.raise_for_status()
+        except HTTPError as e:
+            raise FreshdeskError("{}: {}".format(e, j))
+
+        return j
+
+    def get(self, url: str, params: Mapping = None):
+        """Wrapper around request.get() to use the API prefix. Returns a JSON response."""
+        params = params or {}
+        req = self._session.get(self._api_prefix + url, params=params)
+        return self._parse_and_handle_errors(req)
+
+
+class StreamAPI(ABC):
+    """Basic stream API that allows to iterate over entities"""
+
+    result_return_limit = 100  # maximum value
+
+    def __init__(self, api: API, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._api = api
+
+    @abstractmethod
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        """Iterate over entities"""
+
+    def read(self, getter: Callable, params: Mapping[str, Any] = None) -> Iterator:
+        """Read using getter"""
+        params = params or {}
+        pagination_params = dict(per_page=self.result_return_limit, page=1)
+
+        while True:
+            batch = list(getter(params={**params, **pagination_params}))
+            yield from batch
+
+            if len(batch) < self.result_return_limit:
+                break
+
+            pagination_params["page"] += 1
+
+
+class AgentsAPI(StreamAPI):
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        """Iterate over entities"""
+        yield from self.read(partial(self._api.get, url="agents"))
+
+
+class CompaniesAPI(StreamAPI):
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        """Iterate over entities"""
+        yield from self.read(partial(self._api.get, url="companies"))
+
+
+class ContactsAPI(StreamAPI):
```

Comment:
I know it is a bit repetitive, but it is not hard to read and will be changed during my work on incremental sync

Reply:
... probably